### PR TITLE
Lua API: Log incorrect parameter types as error

### LINKED ISSUE #9950: Change registration type checking to show a warning rather than throw an error
https://forum.minetest.net/viewtopic.php?f=6&t=24817

### LINKED ISSUE #9964: Lack of coercion in new registration type checking
See Lua coercion: https://www.lua.org/manual/2.4/node6.html#coercion

There are some false positives. In 5.2.0, you're able to use a number as the text of a HUD text element. In 5.3.0-dev, it throws an error.

It's perfectly valid for a number to be used, Lua even implicitly converts [it in lua_getstring](https://www.lua.org/manual/2.4/node15.html), this is called type coercion. `lua_isstring` returns true on numbers \o/

_Originally posted by @rubenwardy in https://github.com/minetest/minetest/issues/9950#issuecomment-636065910_

### PR DESCRIPTION
**Pull request by SmallJoker**
Incorrect parameter types are logged as errors, taking coercion into account.
This is a workaround to ensure mod compatibility.
Duplicate warnings are ignored per server instance.

Closes #9950 
Fixes #9964

## To do

This PR is Ready for Review.

## How to test

1) `cd mods`
2) `git clone https://github.com/DokimiCU/aotearoa.git --depth 2`
3) Enable mod & start game
4) Warnings (but not repeating) in the logs

### CODE REVIEW

File: src/script/common/c_converter.cpp [hunk 1]
```diff
@@ -458,11 +459,28 @@ size_t read_stringlist(lua_State *L, int index, std::vector<std::string> *result
 
 bool check_field_or_nil(lua_State *L, int index, int type, const char *fieldname)
 {
-	if (lua_isnil(L, index))
+	static thread_local std::vector<u64> warned_msgs;
```
Comment (sfan5) [inline src/script/common/c_converter.cpp:462]: a set would be more efficient

**Comment by SmallJoker** [comment]
Update as of the most recent commit:

- Coercion fixes (Fixes #9964)
- Logging to errorstream (was warningstream)

**How to test**

5) L431 in `local vscale = tostring(aotearoa.plantlist[i][5])` in `node_defs.lua`

**Comment by SmallJoker** [comment]
http://irc.minetest.net/minetest-dev/2020-06-01#i_5696356

**Comment by paramat** [comment]
This also fixed #9841

**Pull request merged**